src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -0,0 +1,121 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package com.mysql.cj.jdbc.ha.plugins;
+
+import com.mysql.cj.Messages;
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.log.Log;
+
+import com.amazonaws.services.secretsmanager.AWSSecretsManager;
+import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
+import com.amazonaws.services.secretsmanager.model.GetSecretValueRequest;
+import com.amazonaws.services.secretsmanager.model.GetSecretValueResult;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.sql.SQLException;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+public class AWSSecretsManagerPlugin implements IConnectionPlugin {
+  private final ICurrentConnectionProvider currentConnectionProvider;
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+  private final PropertySet propertySet;
+  private final ObjectMapper mapper = new ObjectMapper();
+  private final String region;
+  private final String secretId;
+
+  public AWSSecretsManagerPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+    this.region = propertySet.getStringProperty(PropertyKey.secretsManagerRegion).getValue();
+    this.secretId = propertySet.getStringProperty(PropertyKey.secretsManagerSecretId).getValue();
+  }
+
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    AWSSecretsManager client  = AWSSecretsManagerClientBuilder.standard()
+        .withRegion(this.region)
+        .build();
+    GetSecretValueRequest getSecretValueRequest = new GetSecretValueRequest().withSecretId(this.secretId);
+    GetSecretValueResult getSecretValueResult;
+    String secret = "";
+    Properties updatedProperties = new Properties();
+    updatedProperties.putAll(connectionUrl.getOriginalProperties());
+
+    try {
+      getSecretValueResult = client.getSecretValue(getSecretValueRequest);
+      if (getSecretValueResult.getSecretString() != null) {
+        secret = getSecretValueResult.getSecretString();
+      }
+
+      JsonNode jsonObject = mapper.readTree(secret);
+      updatedProperties.put("username", jsonObject.get("username").asText());
```

Comment:
Should we consider to cache obtained username and pass for some time like 15 mins (configurable by user)?

Reply:
If the secrets are only fetched during openInitialConnection, is caching needed?